tests/basis_test.py

```diff
@@ -3,328 +3,274 @@
-    f0 = ChebyshevFirstKind(0)
-    f1 = ChebyshevFirstKind(1)
-    for i in [-1, -0.5, 0, 0.5, 1]:
-        assert f0(i) == 1
-        assert f1(i) == i
-
-def test_cheb_call_random_points():
+import smolyay
+
+
+def test_cheb_initial():
+    """test degrees returns correctly"""
+    f2 = smolyay.basis.ChebyshevFirstKind(2)
+    assert f2.degree == 2
+    assert isinstance(f2.degree, int)
+    assert numpy.array_equal(f2.domain, [-1, 1])
+    f2.degree = 3
+    assert f2.degree == 3
+    assert isinstance(f2.degree, int)
+
+
+@pytest.mark.parametrize("degree", list(range(3)))
+def test_cheb_call(degree):
     """Test chebyshev polynomial at some degree at some input"""
-    numpy.random.seed(567)
-    ns = numpy.random.randint(20,size = 20)
-    xs = numpy.random.rand(20) * 2 - 1
-    for n,x in zip(ns,xs):
-        f = ChebyshevFirstKind(n)
-        assert numpy.isclose(f(x),special.eval_chebyt(n,x))
-
-def test_cheb_call_random_points_multi_input():
-    """Test that chebyshev polynomial call handles multiple x inputs"""
-    numpy.random.seed(567)
-    xs = numpy.random.rand(20) * 2 - 1
-    f0 = ChebyshevFirstKind(0)
-    f1 = ChebyshevFirstKind(1)
-    fn = ChebyshevFirstKind(5)
-    assert numpy.allclose(f0(xs),special.eval_chebyt(0,xs))
-    assert numpy.allclose(f1(xs),special.eval_chebyt(1,xs))
-    assert numpy.allclose(fn(xs),special.eval_chebyt(5,xs))
+    f = smolyay.basis.ChebyshevFirstKind(degree)
+    assert f(0.5) == pytest.approx(special.eval_chebyt(degree, 0.5))
+    assert f(1) == pytest.approx(special.eval_chebyt(degree, 1))
+    assert f(-1) == pytest.approx(special.eval_chebyt(degree, -1))
 
 
-def test_cheb_derivative():
-    """Test if the correct derivative is generated."""
-    f0 = ChebyshevFirstKind(0)
-    f1 = ChebyshevFirstKind(1)
-    f2 = ChebyshevFirstKind(2)
-    assert f0.derivative(1) == pytest.approx(0)
-    assert f1.derivative(1) == pytest.approx(1)
-    assert f2.derivative(1) == pytest.approx(4)
-    
-    assert f0.derivative(-0.5) == pytest.approx(0)
-    assert f1.derivative(-0.5) == pytest.approx(1)
-    assert f2.derivative(-0.5) == pytest.approx(-2)
-
-    assert numpy.isclose(f0.derivative([1,-0.5]),[0,0]).all()
-    assert numpy.isclose(f1.derivative([1,-0.5]),[1,1]).all()
-    assert numpy.isclose(f2.derivative([1,-0.5]),[4,-2]).all()
-
-
-def test_cheb_2nd_derivative():
-    """Test if the correct derivative is generated."""
-    u0 = ChebyshevSecondKind(0)
-    u1 = ChebyshevSecondKind(1)
-    u2 = ChebyshevSecondKind(2)
-    assert u0.derivative(1) == pytest.approx(0)
-    assert u1.derivative(1) == pytest.approx(2)
-    assert u2.derivative(1) == pytest.approx(8)
+@pytest.mark.parametrize("degree", list(range(3)))
+def test_cheb_call_random_points_multi_input(degree):
+    """Test chebyshev polynomial call handles inputs with complex shape"""
+    f = smolyay.basis.ChebyshevFirstKind(degree)
+    numpy.random.seed(567)
+    xs = numpy.random.rand(20, 3, 4, 6) * 2 - 1
```

Comment:
It would be better not to use random points. You also you don't need so many points along each dimension, but you do want to have more systematic coverage of certain dimension sizes.

I would update this test so that it covers first 2d arrays, including special cases where the array shape is 1 along one or both dimensions.

You can then write a test that covers 3d arrays, including these special cases again. There is probably no need to go to higher dimensionality than that.

You also need a test that covers 1d arrays. It seems like that might have gotten lost when you added the tests on single inputs? That should cover similarly the case where the array has only 1 element and where it has more than one element.

It might be better to make each of these its own test function (1d, 2d, and 3d).

Reply:
Should it also test that the shape is maintained? Like if the input's shape is (1, 5) or (5, 1) or (1, 1, 5, 1, 1, 1)

Reply:
Yes, definitely. The output shape should match the input.

Reply:
Done. There are now tests for scalars, 1D arrays, 2D arrays, and 3D arrays for the call and derivative of each basis function.